File: android/app/src/main/java/com/ayuvo/health/data/metrics/MetricCatalogData.kt
package com.ayuvo.health.data.metrics

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.intOrNull

/**
 * Parsed `shared/metrics/metric_catalog.json` (docs/ui-structure.md §4). Pure Kotlin so the
 * reference port and its vector tests run on the JVM; the app loads the bundled asset copy
 * through [com.ayuvo.health.ui.metrics.MetricCatalog].
 */
data class CatalogDomain(
    val id: String,
    val title: String,
    val titleRes: String,
    val colourHex: String,
    val colourHexDark: String,
    val iconAndroid: String,
    val iconIos: String,
    val browseOrder: Int,
    val target: String
)

data class CatalogUnit(val canonical: String, val decimals: Int, val pref: String?)

data class CatalogMetric(
    val key: String,
    val domain: String,
    val title: String,
    val titleRes: String,
    val iconAndroid: String,
    val iconIos: String,
    val store: String,
    val field: String,
    val unit: CatalogUnit,
    val aggregation: String,
    val chartKind: String,
    val dayBucket: String,
    val ranges: List<String>,
    val goalSource: String,
    val defaultFavouriteOrder: Int?,
    val browseSection: String,
    val browseOrder: Int,
    val about: String
)

data class CatalogOverride(
    val id: String,
    val domain: String?,
    val goalSource: String?,
    val defaultFavouriteOrder: Int?,
    val browseHidden: Boolean,
    /** Optional per-metric icon; null falls back to the domain icon (docs/ui-structure.md §4). */
    val iconAndroid: String? = null,
    val iconIos: String? = null
)

data class CatalogBrowseSection(val id: String, val domain: String, val title: String, val order: Int)

data class MetricCatalogData(
    val domains: List<CatalogDomain>,
    val metrics: List<CatalogMetric>,
    val browseSections: List<CatalogBrowseSection>,
    val categoryDomains: Map<String, String>,
    val aggregationMap: Map<String, String>,
    val chartKindMap: Map<String, String>,
    val overrides: List<CatalogOverride>,
    val macroColours: Map<String, String>,
    val favouritesMax: Int,
    val favouritesPrefKey: String,
    val favouritesLegacyPrefKey: String,
    val dailyStepGoalDefault: Int,
    val dailyStepGoalMin: Int,
    val dailyStepGoalMax: Int
) {
    val domainById: Map<String, CatalogDomain> = domains.associateBy { it.id }
    val metricByKey: Map<String, CatalogMetric> = metrics.associateBy { it.key }
    val overrideById: Map<String, CatalogOverride> = overrides.associateBy { it.id }

    companion object {
        private val json = Json { ignoreUnknownKeys = true }

        fun parse(text: String): MetricCatalogData {
            val root = json.parseToJsonElement(text) as JsonObject
            val health = root.o("health")
            val favourites = root.o("favourites")
            val stepGoal = root.o("prefs").o("dailyStepGoal")
            return MetricCatalogData(
                domains = root.arr("domains").map { d ->
                    d as JsonObject
                    CatalogDomain(
                        id = d.s("id"), title = d.s("title"), titleRes = d.s("title_res"),
                        colourHex = d.s("colour_hex"), colourHexDark = d.s("colour_hex_dark"),
                        iconAndroid = d.o("icon").s("android"), iconIos = d.o("icon").s("ios"), browseOrder = d.i("browse_order") ?: 0,
                        target = d.s("target")
                    )
                },
                metrics = root.arr("metrics").map { m ->
                    m as JsonObject
                    val unit = m.o("unit")
                    val fav = m["default_favourite"] as? JsonObject
                    CatalogMetric(
                        key = m.s("key"), domain = m.s("domain"), title = m.s("title"), titleRes = m.s("title_res"),
                        iconAndroid = m.o("icon").s("android"), iconIos = m.o("icon").s("ios"),
                        store = m.o("source").s("store"), field = m.o("source").s("field"),
                        unit = CatalogUnit(unit.s("canonical"), unit.i("decimals") ?: 0, unit.sOrNull("pref")),
                        aggregation = m.s("aggregation"), chartKind = m.s("chart_kind"), dayBucket = m.s("day_bucket"),
                        ranges = (m["ranges"] as JsonArray).map { (it as JsonPrimitive).content },
                        goalSource = m.s("goal_source"),
                        defaultFavouriteOrder = fav?.let { if (it.b("enabled")) it.i("order") else null },
                        browseSection = m.s("browse_section"), browseOrder = m.i("browse_order") ?: 0,
                        about = m.s("about")
                    )
                },
                browseSections = root.arr("browse_sections").map { b ->
                    b as JsonObject
                    CatalogBrowseSection(b.s("id"), b.s("domain"), b.s("title"), b.i("order") ?: 0)
                },
                categoryDomains = health.o("category_domains").strings(),
                aggregationMap = health.o("aggregation_map").strings(),
                chartKindMap = health.o("chart_kind_map").strings(),
                overrides = health.arr("overrides").map { o ->
                    o as JsonObject
                    val fav = o["default_favourite"] as? JsonObject
                    CatalogOverride(
                        id = o.s("id"), domain = o.sOrNull("domain"), goalSource = o.sOrNull("goal_source"),
                        defaultFavouriteOrder = fav?.let { if (it.b("enabled")) it.i("order") else null },
                        browseHidden = o.b("browse_hidden"),
                        iconAndroid = (o["icon"] as? JsonObject)?.sOrNull("android"),
                        iconIos = (o["icon"] as? JsonObject)?.sOrNull("ios")
                    )
                },
                macroColours = root.o("macro_colours").strings(),
                favouritesMax = favourites.i("max") ?: 12,
                favouritesPrefKey = favourites.s("pref_key"),
                favouritesLegacyPrefKey = favourites.s("legacy_pref_key"),
                dailyStepGoalDefault = stepGoal.i("default") ?: 10_000,
                dailyStepGoalMin = stepGoal.i("min") ?: 1_000,
                dailyStepGoalMax = stepGoal.i("max") ?: 50_000
            )
        }

        private fun JsonObject.o(k: String): JsonObject = this[k] as JsonObject
        private fun JsonObject.arr(k: String): JsonArray = this[k] as JsonArray
        private fun JsonObject.prim(k: String): JsonPrimitive? = (this[k] as? JsonPrimitive)?.takeIf { it !is JsonNull }
        private fun JsonObject.s(k: String): String = prim(k)?.content ?: ""
        private fun JsonObject.sOrNull(k: String): String? = prim(k)?.takeIf { it.isString }?.content
        private fun JsonObject.i(k: String): Int? = prim(k)?.intOrNull
        private fun JsonObject.b(k: String): Boolean = prim(k)?.booleanOrNull ?: false
        private fun JsonObject.strings(): Map<String, String> =
            entries.associate { (k, v: JsonElement) -> k to (v as JsonPrimitive).content }
    }
}

File: ios/calorietracker/Models/Metrics/MetricCatalogData.swift
import Foundation

/// Decoded `metric_catalog.json` (bundled copy of `shared/metrics/metric_catalog.json`,
/// docs/ui-structure.md §4). Pure data; safe to read from any isolation domain.
nonisolated struct MetricCatalogData: Decodable, Sendable {
    struct Icon: Decodable, Sendable {
        let android: String
        let ios: String
    }

    struct Domain: Decodable, Sendable {
        let id: String
        let title: String
        let titleRes: String
        let colourHex: String
        let colourHexDark: String
        let icon: Icon
        let browseOrder: Int
        let target: String
    }

    struct Unit: Decodable, Sendable {
        let canonical: String
        let decimals: Int
        let pref: String?
    }

    struct Favourite: Decodable, Sendable {
        let enabled: Bool
        let order: Int?
    }

    struct Source: Decodable, Sendable {
        let store: String
        let field: String
    }

    struct AppMetric: Decodable, Sendable {
        let key: String
        let domain: String
        let title: String
        let titleRes: String
        let icon: Icon
        let source: Source
        let unit: Unit
        let aggregation: String
        let chartKind: String
        let dayBucket: String
        let ranges: [String]
        let goalSource: String
        let defaultFavourite: Favourite?
        let browseSection: String?
        let browseOrder: Int
        let about: String
    }

    struct Override: Decodable, Sendable {
        let id: String
        let domain: String?
        let goalSource: String?
        let defaultFavourite: Favourite?
        let browseHidden: Bool?
        /// Optional per-metric glyph (docs/ui-structure.md §4 "Icons"); falls back to the domain icon.
        let icon: Icon?
    }

    struct Health: Decodable, Sendable {
        let categoryDomains: [String: String]
        let aggregationMap: [String: String]
        let chartKindMap: [String: String]
        let overrides: [Override]
    }

    struct BrowseSection: Decodable, Sendable {
        let id: String
        let domain: String
        let order: Int
        let title: String
    }

    struct Ring: Decodable, Sendable {
        let id: String
        let metric: String
        let goalSource: String
        let domain: String
        let visibleWhen: String
        let noSourceState: String?
    }

    struct Favourites: Decodable, Sendable {
        let prefKey: String
        let legacyPrefKey: String
        let max: Int
    }

    struct IntPref: Decodable, Sendable {
        let `default`: Int
        let min: Int
        let max: Int
        let step: Int
    }

    struct Prefs: Decodable, Sendable {
        let dailyStepGoal: IntPref
    }

    let catalogVersion: Int
    let domains: [Domain]
    let metrics: [AppMetric]
    let browseSections: [BrowseSection]
    let health: Health
    let macroColours: [String: String]
    let summaryRings: [Ring]
    let favourites: Favourites
    let prefs: Prefs

    static let resourceName = "metric_catalog"

    /// The bundled catalog. A missing or corrupt resource is a build error caught by
    /// `MetricCatalogTests`; at runtime it degrades to an empty catalog instead of crashing.
    static let shared: MetricCatalogData = load() ?? .empty

    static func load() -> MetricCatalogData? {
        for bundle in [Bundle.main, Bundle(for: BundleMarker.self)] {
            if let url = bundle.url(forResource: resourceName, withExtension: "json"),
               let data = try? Data(contentsOf: url),
               let decoded = decode(data) {
                return decoded
            }
        }
        return nil
    }

    static func decode(_ data: Data) -> MetricCatalogData? {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return try? decoder.decode(MetricCatalogData.self, from: data)
    }

    private final class BundleMarker {}

    static let empty = MetricCatalogData(
        catalogVersion: 0, domains: [], metrics: [], browseSections: [],
        health: Health(categoryDomains: [:], aggregationMap: [:], chartKindMap: [:], overrides: []),
        macroColours: [:], summaryRings: [],
        favourites: Favourites(prefKey: "summaryFavourites", legacyPrefKey: "healthHomeTiles", max: 12),
        prefs: Prefs(dailyStepGoal: IntPref(default: 10_000, min: 1_000, max: 50_000, step: 500))
    )

    // MARK: - Lookups

    var appMetricsByKey: [String: AppMetric] {
        Dictionary(metrics.map { ($0.key, $0) }, uniquingKeysWith: { first, _ in first })
    }

    func domain(_ id: String) -> Domain? { domains.first { $0.id == id } }

    func override(_ id: String) -> Override? { health.overrides.first { $0.id == id } }

    /// Default favourites in catalog order (app metrics and health overrides), docs §4.
    var defaultFavourites: [String] {
        var rows: [(Int, String)] = []
        for metric in metrics {
            if let fav = metric.defaultFavourite, fav.enabled, let order = fav.order { rows.append((order, metric.key)) }
        }
        for override in health.overrides {
            if let fav = override.defaultFavourite, fav.enabled, let order = fav.order { rows.append((order, override.id)) }
        }
        return rows.sorted { $0 < $1 }.map(\.1)
    }
}
